src/runtime_src/core/tools/common/ReportAieShim.cpp

```diff
@@ -0,0 +1,338 @@
+/*
+  Copyright (C) 2020 Xilinx, Inc
+ 
+  Licensed under the Apache License, Version 2.0 (the "License"). You may
+  not use this file except in compliance with the License. A copy of the
+  License is located at
+ 
+      http://www.apache.org/licenses/LICENSE-2.0
+ 
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+  License for the specific language governing permissions and limitations
+  under the License.
+ */
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include "ReportAieShim.h"
+#include "core/common/query_requests.h"
+#include "core/common/device.h"
+#include <boost/optional/optional.hpp>
+#include <boost/property_tree/json_parser.hpp>
+
+#define fmt4(x) boost::format("    %-22s: " x "\n")
+#define fmt8(x) boost::format("        %-22s: " x "\n")
+#define fmt12(x) boost::format("            %-22s: " x "\n")
+#define fmt16(x) boost::format("                %-22s: " x "\n")
+
+namespace qr = xrt_core::query;
+
+boost::property_tree::ptree
+populate_aie_shim(const xrt_core::device * device, const std::string& desc)
+{
+  boost::property_tree::ptree pt;
+  std::string aie_data;
+  std::vector<std::string> graph_status;
+  pt.put("description", desc);
+  boost::property_tree::ptree tile_array;
+  boost::property_tree::ptree _pt;
+  boost::property_tree::ptree empty_pt;
+
+  try {
+    aie_data = xrt_core::device_query<qr::aie_shim_info>(device);
+    std::stringstream ss(aie_data);
+    boost::property_tree::read_json(ss, _pt);
+  } catch (const std::exception& ex){
+    pt.put("error_msg", ex.what());
+    return pt;
+  }
+
+  try {
+    for (auto& as: _pt.get_child("aie_shim")) {
+      boost::property_tree::ptree& oshim = as.second;
+      boost::property_tree::ptree ishim;
+      int col = oshim.get<uint32_t>("col");
+      int row = oshim.get<uint32_t>("row");
+      
+      ishim.put("column", col);
+      ishim.put("row", row);
+      int count = 0;
+
+      for (auto& node: oshim.get_child("dma.Channel_status.MM2S", empty_pt)) {
+        ishim.put("dma.channel_status.mm2s.channel_"+std::to_string(count++), node.second.data());
+      }
+
+      count = 0;
+      for (auto& node: oshim.get_child("dma.Channel_status.S2MM", empty_pt)) {
+        ishim.put("dma.channel_status.s2mm.channel_"+std::to_string(count++), node.second.data());
+      }
+
+      count = 0;
+      std::string mode;
+      for (auto& node: oshim.get_child("dma.Mode.MM2S", empty_pt)) {
+        mode = mode+(mode.empty()?"":", ")+node.second.data();

```

Comment:
Wouldn't `+=` be better then `mode = mode + ...`

Reply:
fixed